### PR TITLE
BuildTools 2.0

### PR DESCRIPTION
**Pull request by weshaggard**
Update BuildTools to consume all the changes that we did as part of dev/eng work and left in the corefx and coreclr branches.

These changes have numerous breaking changes for consumers of BuildTools. Some include but not limited to:
- Assumption that you are using msbuild base CLI
- Test infrastructure has changes to depend on a shared runtime
- Signing has changed to use AssemblyKey instead of Use* properties.
- Addition of configuration system

At this point the PR is up to get some early feedback and only contains the task library and the things from Tools-Override. I'm also looking at moving some other common targets with this to avoid more duplication.

### CODE REVIEW

**Comment by weshaggard** [comment]
I've verified locally that corefx can consume this package. To help with testing in other repo's I've uploaded a private build of the BuildTools package with these changes. https://dotnet.myget.org/F/dotnet-buildtools/api/v2/package/Microsoft.DotNet.BuildTools/2.0.0-prerelease-01526-0 Please use that to test out the changes.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets [hunk 1]
```diff
@@ -0,0 +1,522 @@
+<Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (mellinoe) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets]: This should go into the other project, "Microsoft.DotNet.Build.CloudTestTasks".

**Comment by ericstj** [comment]
Looks reasonable.  I take it you've done a diff of the tools folder before and after your change and confirmed no diff in targets/props other than your changes?

Can you show me your changes to make corefx consume this?  I'd like to see how you are hooking up the configuration system.

**Comment by mellinoe** [comment]
Looks reasonable to me, too. Let's get it in as long as corefx builds cleanly with all of this removed from there.

**Comment by weshaggard** [comment]
>Looks reasonable. I take it you've done a diff of the tools folder before and after your change and confirmed no diff in targets/props other than your changes?

I did a diff of the files and merged them from corefx\Tools-Override to BuildTools and did testing.

>Can you show me your changes to make corefx consume this? I'd like to see how you are hooking up the configuration system.

I've got another PR I'm about to submit for corefx.

**Comment by wtgodbe** [comment]
With the changes to packageresolve.targets, will moving away from project.json files be a prerequisite for CoreCLR to consume these changes? Is it documented anywhere what assumptions have changed in that context? That change could be pretty intensive for CoreCLR.

File: src/Microsoft.DotNet.Build.Tasks/GenerateDepsJson.cs [hunk 2]
```diff
@@ -0,0 +1,177 @@
+﻿using System.Collections.Generic;
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Build.Tasks/GenerateDepsJson.cs]: Missing license header

**Comment by weshaggard** [comment]
>With the changes to packageresolve.targets, will moving away from project.json files be a prerequisite for CoreCLR to consume these changes? Is it documented anywhere what assumptions have changed in that context? That change could be pretty intensive for CoreCLR.

Yes that will be a prerequisite. @mellinoe can help you with the necessary changes but hopefully it is mostly as simple as translating the project.json files to ProjectReferences. I'm about to do this exercise for the standard repo as well and so that will give you some general guidance, here is the corefx commit https://github.com/dotnet/corefx/commit/8b471f76eef041f0cf40fec65c99a29b235bbe7d. 

If it turns out to be crazy we can consider supporting both for some time.

File: src/Microsoft.DotNet.Build.Tasks/GenerateDepsJson.cs (hunk 2 above)
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/GenerateDepsJson.cs]: Fixed.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets (hunk 1 above)
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets]: Not sure @MattGal any opinions?

File: src/Microsoft.DotNet.Build.Tasks/Configuration/Configuration.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                    configurationBuilder.Append(ConfigurationFactory.PropertySeperator);
+                }
+                configurationBuilder.Append(value.Value);
+            }
+
+            return configurationBuilder.ToString();
+        }
+
+        /// <summary>
+        /// Get properties assoicated with this configuration
```
Comment (MattGal) [inline src/Microsoft.DotNet.Build.Tasks/Configuration/Configuration.cs]: associated

File: src/Microsoft.DotNet.Build.Tasks/SaveItems.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+{
+    public class SaveItems : BuildTask
+    {
+        [Required]
+        public string ItemName { get; set; }
+
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        [Required]
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/SaveItems.cs]: I added output to this with https://github.com/dotnet/corefx/commit/7389e08300d9bab19e2604e61a30bf0720444c7e

Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/SaveItems.cs]: ah yes I just remerged the targets but didn't do the code. I will do that now.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets (hunk 1 above)
Comment (mellinoe) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets]: I mean that this file already exists in the other folder, unless I'm looking at it wrong.

Comment (MattGal) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets]: Have not had time to read the whole PR yet, but it does make sense going to the CloudTestTasks project if it's still there.

However, if we're revving the version of buildtools, I'd much rather get rid of this file entirely.  CoreFX and CoreClr now use CloudTest.Helix.Targets and anyone adopting a new buildtools can bring their stuff up on the new Targets file in less than a day and have at least 90% fewer headaches when maintaining the file.

File: src/Microsoft.DotNet.Build.Tasks/Configuration/ConfigurationFactory.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        /// <summary>
+        /// Parses a configuration string to return a Configuration.
+        /// </summary>
+        /// <param name="configurationString"></param>
+        /// <returns></returns>
+        internal Configuration ParseConfiguration(string configurationString, bool permitUnknownValues = false)
+        {
+            var values = configurationString.Split(PropertySeperator);
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/Configuration/ConfigurationFactory.cs]: typo, PropertySeparator

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets (hunk 1 above)
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/CloudTest.targets]: OK I removed CloudTest.targets, both this one that I accidentally added and the one under CloudTestTasks.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/crossgen.sh [hunk 6]
```diff
@@ -81,15 +81,19 @@ __toolsDir=$__scriptpath/../Tools
...
-__rid=$($__dotnet --info | sed -n -e 's/^.*RID:[[:space:]]*//p')
 
-if [[ $__rid == *"osx"* ]]; then
-    __packageRid="osx-x64"
-elif [[ $__rid == *"rhel.7"* || $__rid == *"centos.7"* ]]; then
-    __packageRid="rhel.7-x64"
-else
-    __packageRid=$__rid
-fi
+case $(uname -s) in
```
Comment (ellismg) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/crossgen.sh]: Thanks!

**Comment by weshaggard** [comment]
I think this is ready now. As soon as I get a green CI in corefx I'm going to merge and then get the corefx and standard PR updated with the officially build BuildTools package.

**Pull request merged**